### DISCUSSION TITLE
Binding deep nested data broken.

### DISCUSSION THREAD

**Discussion by betaselect** [livewire/livewire#3076] [7. Bugs (archived)]
### Description
First of all, I love this package! It takes a way the hurdle of JS in dynamic interfaces. 

My problem:
I have build a very dynamic form. This is the context: A candidate can have multiple employers and at a employer he can have (had) multiple positions. Therefore we have to bind the form item of a position on the following way: 

`
protected $rules = [
        'employers.*.employer_name' => 'required|string|min:3',
        'employers.*.positions.*.position_title' => 'required',
];
`
The binding of the employer name works. That is also synced between the frontend and the Livewire component. 
However when loading the page the **position title** is not displayed. It stays blank. 

I can conclude that the bind is successful because of two reasons:
- as soon as I enter data in the form field after a few seconds the input disappears.
- when I change the name of the field, an error pops up that I have to bind the input with $rules. 

In my opinion this means that somewhere in the package the data is incorrectly handled. I have tried to figure out the behavior by looking into the hydrate / dehydrate method for dotNotedData. 
However by checking the data via dd() debugging all seems fine. I even see an update object with the original value (which is not displayed in the text input) and the new input.

This supports my hypothesis that somewhere in the source of Livewire the data is not getting correctly placed. However I ran into the limits of my debugging skills and PHP skills.

I hope someone can help me figure this out and find a solution. I am more then willing to submit a PR for this problem however I am unable to find the core issue. 

### Exact steps to reproduce

This is a example of the data structure I use:
```

  0 => array:6 [▼
    "id" => 91
    "employer_name" => "Employer 1"
    "candidate_id" => 33
    "created_at" => "2020-08-24T09:06:39.000000Z"
    "updated_at" => "2020-08-24T09:06:39.000000Z"
    "positions" => []
  ]
  1 => array:6 [▼
    "id" => 92
    "employer_name" => "Employer 2"
    "candidate_id" => 33
    "created_at" => "2020-08-24T09:06:58.000000Z"
    "updated_at" => "2020-08-24T09:06:58.000000Z"
    "positions" => array:2 [▼
      0 => array:11 [▼
        "id" => 70
        "position_title" => "Example position"
        "description" => """
          Test description
          """
        "start_date" => "2016-03-24"
        "end_date" => null
        "date_notation" => "m-Y"
        "is_current_position" => "true"
        "employer_id" => 92
        "candidate_id" => 33
        "created_at" => "2020-08-24T09:06:58.000000Z"
        "updated_at" => "2020-08-24T09:06:58.000000Z"
      ]
      1 => array:11 [▼
        "id" => 71
        "position_title" => "Example position"
        "description" => """
          Testing engineer\r\n
         
          """
        "start_date" => "2006-12-24"
        "end_date" => "2006-03-24"
        "date_notation" => "Y"
        "is_current_position" => null
        "employer_id" => 92
        "candidate_id" => 33
        "created_at" => "2020-08-24T09:06:58.000000Z"
        "updated_at" => "2020-08-24T09:06:58.000000Z"
      ]
    ]
  ]
  2 => array:6 [▼
    "id" => 93
    "employer_name" => "Employer 3 "
    "candidate_id" => 33
    "created_at" => "2020-08-24T09:06:58.000000Z"
    "updated_at" => "2020-08-24T09:06:58.000000Z"
    "positions" => array:1 [▼
      0 => array:11 [▼
        "id" => 72
        "position_title" => "Engineer"
        "description" => ""Description"
        "start_date" => "1992-08-24"
        "end_date" => "2006-08-24"
        "date_notation" => "Y"
        "is_current_position" => null
        "employer_id" => 93
        "candidate_id" => 33
        "created_at" => "2020-08-24T09:06:58.000000Z"
        "updated_at" => "2020-08-24T09:06:58.000000Z"
      ]
    ]
  ]

```

### Stripped-down, copy-pastable code snippets

This is the (simplified) Livewire component i use:

```
namespace App\Http\Livewire\Candidate;

use App\Models\Candidate\CandidateEmployer;
use App\Models\Candidate\CandidatePosition;
use App\Models\Candidate\CandidateResume;
use Livewire\Component;

class ResumeWorkExperienceEdit extends Component
{
public $resume;
public $candidate;
public $employers = [];


protected $rules = [
    'employers.*.employer_name' => 'required|string|min:3',
    'employers.*.positions.*.position_title' => 'required|string|min:3',
];

public function mount()
{
    $this->resume = CandidateResume::where('candidate_id', $this->candidate->id)->first();
    $this->employers = old('education', $this->resume->employers);
}



public function resumeEdited()
{
    $this->emit('resumeEdited');
}

public function render()
{
    return view('livewire.candidate.resume-work-experience-edit');
}
}
```

This is the frontend code:

```
@foreach($employers as $i => $employer)
<div class="{{ !$loop->first ? 'border-t border-gray-200' : ''}} py-4">
   <h3 class="text-md leading-6 font-medium text-gray-900">
      Werkgever {{ $loop->iteration }}
      <button type="button" wire:click.prevent="removeEducation({{ $i }})"
         class="float-right relative inline-flex items-center px-1 py-1 rounded-md border border-gray-300 bg-white text-sm leading-5 font-medium text-gray-700 hover:text-purple-500 focus:z-10 focus:outline-none focus:border-blue-300 focus:shadow-outline-blue active:bg-gray-100 active:text-gray-700 transition ease-in-out duration-150">
         <svg class="h-4 w-4 text-gray-400" x-description="Heroicon name: bookmark"
            xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
            <path
               d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z">
            </path>
         </svg>
      </button>
   </h3>
   <div class="grid grid-cols-6 gap-4">
      <div class="col-span-6">
         <x-form-input name="employers.{{ $i }}.employer_name"
            wire:model="employers.{{ $i }}.employer_name" label="Naam werkgever" />
      </div>
   </div>
   @foreach($employer->positions as $i_position => $position)
   <div class="grid grid-cols-6 gap-4">
      <div class="col-span-6">
         <x-form-input name="employers.{{ $i }}.positions.{{ $i_position }}.position_title"
            wire:model="employers.{{ $i }}.positions.{{ $i_position }}.position_title"
            label="Naam werkgever" />
      </div>
   </div>
   @endforeach
</div>
@endforeach
```

**Context**
- PHP: 7.4.10
- Livewire version: v2.3.0 
- Laravel version: v8.10.0 
- Browser: Safari

**Comment by colinmackinlay** [comment]
Same Issue. I have an ELoquent model for "Area" fetched with a hasMany relationship for "Items" within the area. I need to bind a series of checkboxes to each "Item"s 'active' attribute

If I `wire:model` to `area.items.{{ $index }}.active` and I have `$rules='area.items.*.active' => 'required'` then the checkboxes display initial state correctly. Toggling the box gives the missing validation error here:
```
        $propertyName = $this->beforeFirstDot($name);

 

        throw_if(

            $this->{$propertyName} instanceof Model && $this->missingRuleFor($name),

            new CannotBindToModelDataWithoutValidationRuleException($name, $this::getName())

        );
```
I wonder if the clue is in the 'beforeFirstDot()' method?

I've tried copying the "Items" on mount to their own local `$items` variable and binding to that with appropriate rules but it won't display the initial state even though the variable is definitely correctly initialised with the collection. I can click on the checkbox but the binding doesn't work and the state does not persist with the box toggling back.

Confused.com!

**Comment by basepack** [comment]
I also have a related issue, cannot toggle a boolean on a Eloquent model property:

```php
// Eloquent model
protected $casts = [
    'a_boolean' => 'boolean',
];
// even setting the attribute by default did not help
protected $attributes = [
   'a_boolean' => false,
];
```

```php
// Component
public $rules = [
   'model.a_boolean => 'boolean'
];

// Edit: I new up a model like this
public function mount(){
    $this->model = Model::make();
}
```

```blade
<div wire:click="$toggle('model.a_boolean')">Toggle It!</div>
@json($model->a_boolean)
```

Click that button and get this error:
```
Livewire\Exceptions\PropertyNotFoundException
Property [$model.a_boolean] not found on component
```

**Comment by uteq** [comment]
I did not try any of the given examples. But my experience with these kind of problems is that whenever working with nested data it is better to create 'child' components. So in your case it would be wise to have a separate Employers component, etc.

**Comment by basepack** [comment]
@uteq but then I have for the form I am working on now at least 3 child components and then need to handle all events which is also not a big deal.


At the moment, this does not work either:

Eloquent Model:
```php
    protected $casts = [
        'shipping_costs'                => 'array', // json
];
```

Livewire Component:
```php
    public $rules = [
        'shippingRule.shipping_costs.*.from' => 'integer',
    ];
```

View:
```blade
<div>
    @foreach($shippingRule->shipping_costs as $key => $data)

        <!-- note: from is an integer in $rules -->
        <x-input.text wire:model="shippingRule.shipping_costs.{{ $key }}.from"/> 
     @endforeach
</div>
```

Now when I update `from` it is saved as an string instead of an integer like in the rules.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by basepack** [comment]
REOPEN

**Comment by pkeogan** [comment]
anyone ever get this working? I am also unable to bind deeply nested data.

Going from a `$collection` of models. into either a casted `$array` or a `$collection` of models that have a relationship gives an error of `Indirect modification of overloaded element of App\Models\Test has no effect
`

**Comment by basepack** [comment]
@joshhanley thanks for the merge. But still my example toggling a model boolean property would not work, with an Eloquent model like so: https://github.com/livewire/livewire/issues/1886#issuecomment-715872847

Maybe it has something to do that I use underscores in the boolean property?

Edit: the exact error I get is this
![stacktrace](https://user-images.githubusercontent.com/939500/103877845-e7ac0a80-50d5-11eb-87ee-22f87dcef43b.png)

**Comment by joshhanley** [comment]
@basepack the update hasn't been tagged yet, can you test using the master branch if you get a chance and let me know if it's working? I just tested it locally with a user that has an `is_active` property using dev-master and it worked for me.

```bash 
composer require livewire/livewire:dev-master
```

Hope this helps!

**Comment by basepack** [comment]
@joshhanley Ah thanks, I did not know it was not tagged yet. I saw that the PR was merged so I was concerned I found a related unfixed bug.

**Comment by joshhanley** [comment]
@basepack no worries! It has been tagged now as 2.3.6. Can you please test it and let us know how it goes?

**Comment by basepack** [comment]
@joshhanley yep it works with 2.3.6, great!

**Comment by joshhanley** [comment]
@basepack awesome! Thanks for testing.

**Comment by CoffeeKing68** [comment]
Hi all

I think I have found an issue that is closely related to this one.

I have a collection of models, `$contents` that each have a collection of models `$tasks` (This example should work with the classic `$user->posts`).

I also have:
- `$parents` where each has an array of `$children`
- `$arrayContents` which = `$contents->toArray()`

If I try to bind to "contents.*.tasks.*.name", and the results are strange.

This is my code.

Test.php
```php
<?php

namespace App\Http\Livewire;

use App\FatContent;
use Livewire\Component;

class Test extends Component
{
    public $parent = [
        0 => [
            "name" => "Parent 1",
            "children" => [
                [
                    "name" => "Child 1"
                ], 
                [
                    "name" => "Child 2"
                ]
            ]
        ],
        1 => [
            "name" => "Parent 2",
            "children" => [
                [
                    "name" => "Child 3"
                ]
            ]
        ]
    ];
    
    public $contents;
    public $arrayContents;
    
    public function mount() {
        $this->contents = FatContent::where('profile_id', 4)
            ->with('tasks')->get();
        
        $this->arrayContents = $this->contents->toArray();
    }

    public function render()
    {
        return view('livewire.test')->extends('app')->section('main-content');
    }
    
    public function rules() {
        return [
            'contents.*.tasks.*.name' => "required",
            "contents.*.number" => "required",
            
            "arrayContents.*.tasks.*.name" => "required"
        ];
    }
}
```

test.blade.php
```blade
<div>
    <div class="border bg-green-500">
        <label for="">parent.0.children.0.name</label>
        <input class="border" type="text" wire:model.lazy="parent.0.children.0.name">
        <div> 
            @json($parent[0]['children'][0]['name'])
        </div>
    </div>
    
    <div class="border bg-green-500">
        <label for="">parent.0.children.1.name</label>
        <input class="border" type="text" wire:model.lazy="parent.0.children.1.name">
        <div> 
            @json($parent[0]['children'][1]['name'])
        </div>
    </div>
    
    <hr>

    <div class="border bg-red-500">
        <label for="">contents.0.tasks.0.name</label>
        <input class="border" type="text" wire:model.lazy="contents.0.tasks.0.name">
        <div> 
            @json($contents[0]['tasks'][0]['name'])
        </div>
    </div>
    
    <div class="border bg-red-500">
        <label for="">contents.0.tasks.1.name</label>
        <input class="border" type="text" wire:model.lazy="contents.0.tasks.1.name">
        <div> 
            @json($contents[0]['tasks'][1]['name'])
        </div>
    </div>
    
    <hr>
    
    <div class="border bg-green-500">
        <label for="">contents.0.number</label>
        <input class="border" type="text" wire:model.lazy="contents.0.number">
        <div> 
            @json($contents[0]['number'])
        </div>
    </div>
    
    <hr>
    
    <div class="border bg-green-500">
        <label for="">arrayContents.0.tasks.0.name</label>
        <input class="border" type="text" wire:model.lazy="arrayContents.0.tasks.0.name">
        <div> 
            @json($arrayContents[0]['tasks'][0]['name'])
        </div>
    </div>
</div>
```

My results:
1. Binding 2 levels deep on `$parents` works
2. Binding 2 levels deep on `$arrayContents` works
3. Binding 1 level deep on `$contents` works
4. Binding 2 levels deep on `$contents` will bind for 1 field, but set other fields to empty arrays

And results 1-3 will set the fields of `$contents` to empty arrays.

Should I open a new issue rather?

Thanks
CoffeeKing

**Comment by joshhanley** [comment]
@CoffeeKing68 nah that is definitely related, so this issue is fine. Livewire currently doesn't support deep nesting so `contents.*.tasks.*.name` won't work.

I'm currently working on deep nesting support, which I have about half finished. I just haven't had the time the last month or so to finish it off.

Hope this helps!

**Comment by CoffeeKing68** [comment]
Thank you @joshhanley , I appreciate your efforts.

However, the deep nesting works for arrays. Do you know why that is? I will probably use that as a proxy for the eloquent collection so my scenario works

Thanks

**Comment by joshhanley** [comment]
@CoffeeKing68 oh yeah arrays are fine. Sorry should have been more clear that it is eloquent models/collections that are limited.

So the reason it's limited is, when you bind to models or eloquent collections, you have to specify in the rules what data can be accessed by the front end. 

But for that to actually work, Livewire iterates through the model/collection properties and only collects and sends to the front end any that are in that list (acting as a whitelist), which is a security feature to prevent all data, such as a user's password field, being sent to the frontend automatically. Instead the developer needs to be deliberate about it. 

The iteration code currently only supports shallow nesting as it is manually specified to a certain depth. To change that to deep nesting requires changing that code to a recursive function then making sure edge cases are covered and ensuring that there isn't excessive data sent to the front end.

I have a proof of concept working with the recursive function that supports something like `posts.*.comments.*.replies.message` for example, but it currently breaks if I do something like `posts.*.comments.*.replies.author.name`. Hopefully will find the time soon to finish it off 🙂

**Comment by CoffeeKing68** [comment]
I see, well I have been using an array as a proxy of object now and it has been working great. 

Thanks for your work Josh

**Comment by egknight** [comment]
Is there any ETA on this? Looking at taking a design decision to either wait on this, or move to arrays.

thanks

**Comment by basepack** [comment]
@egknight well, I would wait for Livewire 3 or just use arrays for the time being. Livewire 3 will potentially introduce its own set of bugs, so using arrays is your safest option for now I think.

**Comment by pkeogan** [comment]
Just want to through out some thoughts on this item, as I was in the same boat for quite some time. 

Recently I decided to refactor my application into a V2, and when I did, any Models that had arrays, were broken down into sub models. So I no longer have a single that has a property or attribute that is a array/json.

for example, I went from the following design
```
$user->settings; // array() = ["foo" => true, "bar" => 10];
```
to
```
$user->settings // $this->hasMany(Setting::class) Setting[name=>'foo", 'value'=>true]
```

Also, within `livewire` I now only allow a single component to mange a single model. So with the above example I would have a livewire component manage by `user` model, then their related `setting` models would have their own component.

This design process will solve the issue most people are having. A hack around is the following
```
class EditForm extends Component
{

    public JobPerformanceEvaluationType $jpeType;
    public JobPerformanceEvaluationTypeVersion $jpeTypeVersion;
    public Collection $jpeTypeObjectives;

    // https://laravel.com/docs/8.x/validation#available-validation-rules
    protected $rules = [
        'jpeTypeVersion.title' => 'required|string|max:500',
        'jpeTypeVersion.description' => 'required|string|max:5000',
        'jpeTypeVersion.instruction' => 'required|string|max:5000',
        'jpeTypeObjectives.*.description' => 'nullable|string',
        'jpeTypeObjectives.*.overall_time_limit_in_mmss' => 'string|nullable',
        'jpeTypeObjectives.*.relative_time_limit_in_mmss' => 'required_with:jpeTypeObjectives.*.parent_id|string|nullable',
        'jpeTypeObjectives.*.parent_id' => 'required_with:jpeTypeObjectives.*.relative_time_limit_in_mmss|numeric|nullable'
    ];

    public function publish()
    {
        $this->saveJpeTypeObjectivesAndReload();
        $this->jpeTypeVersion->published = 1;
        $this->jpeTypeVersion->save();
        return redirect()->route('backend.jpe-type.index');
    }

    public function saveJpeTypeObjectives()
    {
        $this->validateOnly('jpeTypeObjectives.*.description');
        $this->validateOnly('jpeTypeObjectives.*.overall_time_limit_in_mmss');
        $this->validateOnly('jpeTypeObjectives.*.relative_time_limit_in_mmss');
        $this->validateOnly('jpeTypeObjectives.*.parent_id');
        $this->saveJpeTypeObjectivesAndReload();
    }

    public function updating($name, $value)
    {
        if($value == "null"){
            $value = null;
        }

        if( isset(explode('.', $name)[2]) && explode('.', $name)[0] . '.*.' . explode('.', $name)[2] == 'jpeTypeObjectives.*.parent_id')
        {
            $this->jpeTypeObjectives[explode('.', $name)[1]]->parent_id = $value;
            $this->jpeTypeObjectives[explode('.', $name)[1]]->save();
        }
        
        if( isset(explode('.', $name)[2]) && explode('.', $name)[0] . '.*.' . explode('.', $name)[2] == 'jpeTypeObjectives.*.relative_time_limit_in_mmss')
        {
            $this->jpeTypeObjectives[explode('.', $name)[1]]->relative_time_limit_in_mmss = $value;
            $this->jpeTypeObjectives[explode('.', $name)[1]]->save();
        }
        if( isset(explode('.', $name)[2]) && explode('.', $name)[0] . '.*.' . explode('.', $name)[2] == 'jpeTypeObjectives.*.overall_time_limit_in_mmss')
        {
            $this->jpeTypeObjectives[explode('.', $name)[1]]->overall_time_limit_in_mmss = $value;
            $this->jpeTypeObjectives[explode('.', $name)[1]]->save();
        }
        if( isset(explode('.', $name)[2]) && explode('.', $name)[0] . '.*.' . explode('.', $name)[2] == 'jpeTypeObjectives.*.description')
        {
            $this->jpeTypeObjectives[explode('.', $name)[1]]->description = $value;
        }

        if(explode('.', $name)[0] == 'jpeTypeObjectives'){
         //   $this->jpeTypeObjectives[explode('.', $name)[1]]->save();
        }
    }

    public function Mount($jpeType, $jpeTypeVersion)
    {
        $this->jpeType = $jpeType;
        $this->jpeTypeVersion = $jpeTypeVersion;
        $this->jpeTypeObjectives = $this->jpeTypeVersion->jobPerformanceEvaluationTypeObjectives;
    } 

```

you can see to manage the deeply nested data, you need to manage some hacky code. I broke this down into nested compents and have solved my problems, and performance increase is notable. 

TLDR: If your need livewire to access deep into data, your structure of you model is wrong.

**Comment by basepack** [comment]
@pkeogan Thanks for this insight! However how would you go if you have a huge form with 4+ related models which needs to be saved all at once with one save button? At the moment I use your last example with a custom save function for each related model.

**Comment by joshhanley** [comment]
So I just realised I didn't update this thread! Deep nesting has been merged in PR #3409 

It will be available in the next release 🙂

Hope this helps!

**Reply by sawirricardo** [reply]
Hi @joshhanley ,really thank you for your hardwork. However, I think the issue still persists.

Here's mine
user has many userAddresses with property line_1
```
$rules = [
'user.userAddresses.*.line_1' => ['sometimes','nullable','string']
]
```

```
<textarea wire:model='user.userAddresses.0.line_1'></textarea>
```
it doesn't work. However if I output the {{ $user->userAddresses->first()->line_1 }}, the output works, but instantly replaced by livewire (may be due to wire:model?)

Mind you, I also have other heavy array properties. may be due to limited memory? Though I didn't see any memory error 🤔

Thank you @joshhanley

**Reply by joshhanley** [reply]
@sawirricardo well this issue was because deep nesting support didn't exist at all 😆 

If you are experiencing an issue with it now that we have support for deep nesting, can I suggest opening up your own discussion with details of the problem and code samples that reproduce the issue so we can investigate it? 🙂

Thanks!

**Comment by basepack** [comment]
@joshhanley I am so grateful for that PR, no words...

**Reply by joshhanley** [reply]
No worries! 😁
